Atru/mcp server 1.0 ga (#2566)
hashicorp/web-unified-docs: content/terraform-mcp-server/v1.0.x/docs/mcp-server/reference.mdx

---
page_title: Terraform MCP server reference
description: Learn about the configuration options and tools available for the Terraform MCP server.
# START AUTO GENERATED METADATA, DO NOT EDIT
created_at: 2026-01-22T23:00:02-07:00
last_modified: 2026-01-28T10:45:26-07:00
# END AUTO GENERATED METADATA
---

# Terraform MCP server reference

This page contains reference information about the Terraform MCP server, including configuration options and tools.

## Available tools

> The Terraform MCP server provides specialized tools that AI models can use to access current Terraform registry information. These tools work automatically when you ask relevant questions—you don't need to invoke them manually.

Reviewer: ```suggestion
The Terraform MCP server provides specialized tools that AI models can use to access current Terraform registry information. These tools work automatically when you ask relevant questions, you don't need to invoke them manually.
```

## Toolsets 

The server includes the following toolsets:

| Toolset | Description |
| --- | --- |
| `registry` | Tools for accessing the public registry |
| `registry-private` | Tools for accessing private registries |
| `terraform` | Tools for interacting with HCP Terraform and Terraform Enterprise |

You can add the `--toolsets` flag to enable or disable these tools when starting the binary.

> 

Reviewer: ```suggestion
```

Extra newline

Discover and explore community and verified modules from the Terraform registry.

| Tool | Purpose | What it returns |
|------|---------|-----------------|
| `search_modules` | Find modules by name or functionality | Module details including names, descriptions, download counts, and verification status |
| `get_module_details` | Get comprehensive module information | Complete documentation with inputs, outputs, examples, and submodules |
| `get_latest_module_version` | Retrieve the latest version of a specifc module | The latest version of a module |

Example prompt: "Show me modules for deploying a Kubernetes cluster on AWS."

### Policy tools

Access Sentinel policies for governance and compliance requirements.

| Tool | Purpose | What it returns |
|------|---------|-----------------|
| `search_policies` | Find Sentinel policies by topic or requirement | Policy listings with IDs, names, and download statistics |
| `get_policy_details` | Retrieve detailed policy documentation | Policy implementation details and usage instructions |

Example prompt: "What Sentinel policies are available for enforcing security best practices?"

### Terraform Cloud/Enterprise tools

> Manage Terraform Cloud/Enterprise resources including workspaces, runs, variables, and private modules. Some tools require the `ENABLE_TF_OPERATIONS` environment variable for destructive operations.

Reviewer: ```suggestion
Manage HCP Terraform or Terraform Enterprise resources including workspaces, runs, variables, and private modules. Some tools require the `ENABLE_TF_OPERATIONS` environment variable for destructive operations.
```

| Tool | Purpose | What it returns |
|------|---------|-----------------|
| `list_terraform_orgs` | Fetch all Terraform organizations | List of organizations with their details |
| `list_terraform_projects` | Fetch all Terraform projects | List of projects with their metadata |
| `list_workspaces` | Search and list workspaces in an organization | Workspace details including configuration and status |
| `get_workspace_details` | Get comprehensive workspace information | Complete workspace configuration, variables, and state information |
| `create_workspace` | Create a new Terraform workspace | Confirmation of workspace creation (destructive operation) |
| `update_workspace` | Update workspace configuration | Updated workspace configuration (potentially destructive) |

> | `delete_workspace_safely` | Delete workspace if it manages no resources | Confirmation of deletion (requires `ENABLE_TF_OPERATIONS`) |

Reviewer: ```
| `create_workspace` | Create a new Terraform workspace | Confirmation of workspace creation (destructive operation) |
| `update_workspace` | Update workspace configuration | Updated workspace configuration (potentially destructive) |
| `delete_workspace_safely` | Delete workspace if it manages no resources | Confirmation of deletion (requires `ENABLE_TF_OPERATIONS`) |
```

I don't mind that there's not a dedicated column for this information, but are all three of these entries specifying that `ENABLE_TF_OPERATIONS` must be true, just in different ways? Might be a good opportunity to make this a bit more consistent.

| `get_private_module_details` | Get detailed private module information | Complete module details including inputs, outputs, and examples |
| `search_private_providers` | Find private providers in your organization | List of private providers matching search criteria |
| `get_private_provider_details` | Get detailed private provider information | Provider details including usage and version information |
| `get_workspace_policy_sets` | Get the policy sets attached to a particular workspace | A list of policy sets |
| `attach_policy_set_to_workspace` | Attach a policy set to a workspace | Confirmation of attachment |
| `get_token_permissions` | Get the permissions for the `TFE_TOKEN` | A list of actions you can take |
| `get_plan_json_output` | Retrieves the structured JSON output of a Terraform plan, providing detailed resource changes in a machine-readable format that is easier to parse than plain logs |
| `get_plan_details` | Fetches detailed metadata about a specific Terraform plan |
| `get_plan_logs` | Retrieves the execution logs of a specific Terraform plan |
| `get_apply_details` | Fetches detailed metadata about a specific Terraform apply |

> | `get_apply_logs` | Retrieves the execution logs of a specific Terraform apply |

Reviewer: These final entries are missing entries in the "What it returns" column, though I suspect this is on purpose as this is also covered in the "Purpose" column. Your call if you want to add these last few cells.

- "List all private modules related to AWS"
- "Create a variable set for database configuration"

### Stacks tools

Manage Terraform Stacks in an organization.

| Tool | Purpose | What it returns |
|------|---------|-----------------|
| `list_stacks` | Retrieve list of stacks | List of stacks with high level summary |
| `get_stack_details` | Read full details for a specific stack | Full stack configuration |

Example prompts:
- "List all the stacks in my organization"
- "Show me the details for the `dev-environment` stack"

## Available resources

The Terraform MCP server provides several static read-only guides to the MCP clients to retrieve structured, contextual data. It helps generate standardized Terraform code. 

| Resource URI | Kind | Description |
|--------------|------|-------------|
| `/terraform/style-guide` | Resource | Terraform Style Guide - Provides access to the official Terraform style guide documentation in markdown format |
| `/terraform/module-development` | Resource | Terraform Module Development Guide - Comprehensive guide covering module composition, structure, providers, publishing, and refactoring best practices |

> | `/terraform/providers/{namespace}/name/{name}/version/{version}` | Resource Template | Provider Resource Template - Dynamically retrieves detailed documentation and overview for any Terraform provider by namespace, name, and version |

Reviewer: ```suggestion
| `/terraform/style-guide` | Resource | Provides access to the official Terraform style guide documentation in markdown format |
| `/terraform/module-development` | Resource | Comprehensive guide covering module composition, structure, providers, publishing, and refactoring best practices |
| `/terraform/providers/{namespace}/name/{name}/version/{version}` | Resource Template | Dynamically retrieves detailed documentation and overview for any Terraform provider by namespace, name, and version |
```

Your call if you want to keep these, but felt a little redundant IMO

## Environment variables

You can set the following environment variables to configure the server behavior.

| Variable                       | Description                                                                  | Default                      |
|--------------------------------|------------------------------------------------------------------------------|------------------------------|
| `TFE_ADDRESS`                  | HCP Terraform or Terraform Enterprise address                                | `"https://app.terraform.io"` |
| `TFE_TOKEN`                    | Terraform Enterprise API token                                               | `""` (empty)                 |
| `TFE_SKIP_TLS_VERIFY`          | Skip HCP Terraform or Terraform Enterprise TLS verification                  | `false`                      |
| `TRANSPORT_MODE`               | Set to `streamable-http` to enable HTTP transport (legacy `http`)            | `stdio`                      |

> | `TRANSPORT_HOST`               | Host to bind the HTTP server                                                 | `127.0.0.1`                  |

Reviewer: ```suggestion
| `TRANSPORT_HOST`               | Host to bind the HTTP server. Set to `0.0.0.0` to bind to all interfaces                                                 | `127.0.0.1`                  |
```

What do you think about this clarification? I don't believe this is called out anywhere else in the docs

| `OTEL_METRICS_ENABLED`         | Enable tools metrics using OpenTelemtry                                      | `false`                      |
| `OTEL_METRICS_SERVICE_NAME`    | Identifies the source of the metrics                                         | `terraform-mcp-server`       |
| `OTEL_METRICS_EXPORT_INTERVAL` | Controls the frequency of metric flushes                                     | `2`                          |
| `OTEL_METRICS_ENDPOINT`        | OpenTelemetry collector endpoint                                             | `localhost:4318`             |
| `LOG_LEVEL`                    | Logging level: `trace`, `debug`, `info`, `warn`, `error`, `fatal`, `panic`   | `info`                       |
| `LOG_FORMAT`                   | Logging format: `text` or `json` (overrides `--log-format` flag)             | `text`                       |

## Run type options

The `create_run` tool supports the following run types by default:

- `plan_and_apply`: Creates a plan and applies it if approved
- `refresh_state`: Refreshes the state without making changes
- `plan_only`: Creates a plan without applying
- `allow_empty_apply`: Allows applying when no changes are detected

> When `ENABLE_TF_OPERATIONS` is set to `true`, the following additional options become available:

Reviewer: ```suggestion
When you set `ENABLE_TF_OPERATIONS` to `true`, the following additional options become available:
```

- `auto_approve`: Automatically approves and applies the plan
- `is_destroy`: Creates a destroy plan to remove infrastructure

### Destructive operations

Several tools perform destructive operations that can modify or delete infrastructure resources. These tools are disabled by default and require setting `ENABLE_TF_OPERATIONS=true`:

- `action_run`: Can apply or destroy infrastructure
- `delete_workspace_safely`: Permanently deletes workspaces
- Additional run type options in `create_run`

<Warning>

When enabled, use these tools with caution as they can permanently modify or destroy infrastructure resources.

</Warning>

## Telemetry metrics

> When enabled, Terraform MCP server collects the following metrics:

Reviewer: ```suggestion
When enabled, the Terraform MCP server collects the following metrics:
```

| Type | Emitted metrics | Notes |
| --- | --- | --- |

> | Standard HTTP server metrics | <ul><li>`http.server.request.body.size`</li><li>`http.server.response.body.size`</li><li> `http.server.request.duration`</li></ul> |  The server wraps the HTTP mux with `otelhttp.NewHandler(...)` to emit the metrics. |

Reviewer: > The server wraps the HTTP mux with `otelhttp.NewHandler(...)` to emit the metrics.

This is a little outside my wheelhouse, but is this relevant to the user? IMO this reads more like a note that someone working on the server code itself would be concerned with. Would something like "The server emits these metrics using OpenTelemetry" be more relevant to the end user?

Author: Yeah, I'm not sure how in the weeds we need to get, so we've kept this as is. Maybe it's helpful for people making custom integrations.

Reviewer 2: Yeah this HTTP mux info isn't needed for end users of this, I think this is copypasta from the opentelemetry package. `Standard HTTP Metrics` will probably suffice – they'll know we mean request duration, size etc.
